keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,891 @@
+
+If a Bucket is manually deleted by an admin, without deleting the corresponding BucketClaim, then the cluster is considered to be in an invalid state. Manual recovery is possible if data is not already lost. 
+
+# COSI API Reference
+
+## Bucket
+
+Resource to represent a Bucket in OSP. Buckets are cluster-scoped.
+
+```yaml
+Bucket {
+  TypeMeta
+  ObjectMeta
+
+  Spec BucketSpec {
+    // Provisioner is the name of driver associated with this bucket
+    Provisioner string
+
+    // DeletionPolicy is used to specify how COSI should handle deletion of this
+    // bucket. There are 3 possible values:
+    //  - Retain: Indicates that the bucket should not be deleted from the OSP (default)
+    //  - Delete: Indicates that the bucket should be deleted from the OSP
+    //        once all the workloads accessing this bucket are done
+    // +optional
+    DeletionPolicy DeletionPolicy
+
+    // Name of the BucketClass specified in the BucketRequest
+    BucketClassName  string
+
+    // Name of the BucketClaim that resulted in the creation of this Bucket
+    // Optional in case Bucket was created without a BucketClaim i.e. Imported 
+    // +optional
+    BucketClaim corev1.ObjectReference
+
+    // Protocols are the set of data APIs this bucket is expected to support.
+    // The possible values for protocol are:
+    // -  S3: Indicates Amazon S3 protocol
+    // -  Azure: Indicates Microsoft Azure BlobStore protocol
+    // -  GCS: Indicates Google Cloud Storage protocol
+    Protocols []Protocol
+
+    // Parameters is an opaque map for passing in configuration to a driver
+    // for creating the bucket
+    // +optional
+    Parameters map[string]string
+
+    // ExistingBucketID is the unique id of the bucket in the OSP. This field should be
+    // used to specify a bucket that has been created outside of COSI.
+    // This field will be empty when the Bucket is dynamically provisioned by COSI.
+    // +optional
+    ExistingBucketID string
+  }
+
+  Status BucketStatus {
+    // BucketReady is a boolean condition to reflect the successful creation
+    // of a bucket.
+    BucketReady bool
+
+    // BucketID is the unique id of the bucket in the OSP. This field will be
+    // populated by COSI.
+    // +optional
+    BucketID string
+  }
+}
 ```
-message ProvisionerCreateBucketRequest {    
-    // This field is REQUIRED
-    string bucket_name = 1;
 
-    map<string,string> bucket_context = 2;
+## BucketClaim
+
+A claim to create Bucket. BucketClaim is namespace-scoped
+
+```yaml
+BucketClaim {
+  TypeMeta
+  ObjectMeta
 
-    enum AnonymousBucketAccessMode {
-	PRIVATE = 0;
-	PUBLIC_READ_ONLY = 1;
-	PUBLIC_WRITE_ONLY = 2;
-	PUBLIC_READ_WRITE = 3;
-    }
+  Spec BucketClaimSpec {
+    // Name of the BucketClass
+    BucketClassName string
     
-    AnonymousBucketAccessMode anonymous_bucket_access_mode = 3;
-}
+    // Name of a bucket object that was manually 
+    // created to import a bucket created outside of COSI
```

Comment:
Add a comment here, if unspecified, then a new bucket will be dynamically provisioned.